### PR TITLE
core: rpmb fs uses mempool for temporary transfer buffers

### PR DESCRIPTION
**Pull request by etienne-lms**
RPMB FS driver may allocates a temporary buffer of size the one provided
by userland caller. These may be big buffer of dozens of kbytes and may
exhaust the heap capacities. Change the implementation to use a 4kByte
temporary buffer to update RPMB data instead of a allocated buffer of
the object target size.

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

### CODE REVIEW

File: core/tee/tee_rpmb_fs.c [hunk 1]
```diff
@@ -2433,35 +2442,87 @@ static TEE_Result rpmb_fs_write_primitive(struct rpmb_file_handle *fh,
 		 * File must be extended, or update cannot be atomic: allocate,
 		 * read, update, write.
 		 */
+		uintptr_t old_fat = fh->fat_entry.start_address;
```
Comment (jenswi-linaro) [inline core/tee/tee_rpmb_fs.c:2445]: The code would be easier to follow if most of what's in this else was moved into a helper function.

Comment (etienne-lms) [reply core/tee/tee_rpmb_fs.c:2445]: lot of variable to track the 2 sources and 1 dest :(

Comment (jenswi-linaro) [reply core/tee/tee_rpmb_fs.c:2445]:
``` C
static TEE_Result update_write_helper(struct rpmb_file_handle *fh,
                                      size_t pos, const void *buf,
                                      size_t size, uintptr_t new_fat,
                                      size_t new_size)
```
seems manageable.

Comment (etienne-lms) [reply core/tee/tee_rpmb_fs.c:2445]: oh, i single helper function. yes.
sorry, I misread your comment. will do.

**Comment by etienne-lms** [comment]
I rebased and squashed fixes to get the P-R readable.

File: core/tee/tee_rpmb_fs.c [hunk 2]
```diff
@@ -2369,19 +2372,87 @@ static TEE_Result rpmb_fs_read(struct tee_file_handle *tfh, size_t pos,
...
+	size_t rpmb_bytes_to_write = new_size;
+
+	if (!rpmb_bytes_to_write)
+		return TEE_SUCCESS;
+
+	blk_buf = mempool_alloc(mempool_default, TMP_BLOCK_SIZE);
+	if (!blk_buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	while (rpmb_bytes_to_write) {
```
Comment (jenswi-linaro) [inline core/tee/tee_rpmb_fs.c:2397]: I'd prefer dropping `rpmb_bytes_to_write` and test `blk_offset < new_size` instead.

File: core/tee/tee_rpmb_fs.c [hunk 3]
```diff
@@ -2369,19 +2372,87 @@ static TEE_Result rpmb_fs_read(struct tee_file_handle *tfh, size_t pos,
...
+		blk_size = MIN(TMP_BLOCK_SIZE, rpmb_bytes_to_write);
```
Comment (jenswi-linaro) [inline core/tee/tee_rpmb_fs.c:2398]: I'd prefer replacing `rpmb_bytes_to_write` with `new_size - blk_offset`.

**Review by jenswi-linaro — COMMENTED** [review]
With or without my comments addressed.
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/tee/tee_rpmb_fs.c (hunk 2 above)
Comment (etienne-lms) [reply core/tee/tee_rpmb_fs.c:2397]: makes sense. ok.

File: core/tee/tee_rpmb_fs.c [hunk 4]
```diff
@@ -2369,19 +2372,87 @@ static TEE_Result rpmb_fs_read(struct tee_file_handle *tfh, size_t pos,
...
+	size_t old_size = fh->fat_entry.data_size;
+	const uint8_t *rem_buf = buf;
+	size_t rem_size = size;
+	uint8_t *blk_buf = NULL;
+	size_t blk_offset = 0;
+	size_t blk_size = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t rpmb_bytes_to_write = new_size;
+
+	if (!rpmb_bytes_to_write)
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/tee/tee_rpmb_fs.c:2390]: i can drop this test and default init 	`res = TEE_SUCCESS;`

**Comment by jenswi-linaro** [comment]
Looks good.

**Comment by etienne-lms** [comment]
squashed & applied review tag

**Pull request merged**